R/paginate.R

```diff
@@ -0,0 +1,336 @@
+#' Pagination
+#'
+#' @inheritParams req_perform
+#' @param next_request A callback function that takes a two arguments (the
+#'   original request and the response) and returns:
+#'
+#'   * a new [request] to request the next page or
+#'   * `NULL` if there is no next page.
+#' @param page_size A parameter object that specifies how the page size is added
+#'   to the request.
+#' @param total A character that specifies the path where in the body the field
+#'   with the total number of elements is stored.
+#' @param next_url A character that specifies the path where in the body the field
+#'   with the next url of the next page is stored.
+#' @param offset A parameter object that specifies how the offset is added to
+#'   the request.
+#' @param token_field A parameter object that specifies how the next token is
+#'   added to the request.
+#' @param next_token_field A character that specifies the path where in the body
+#'   the field with the token of the next page is stored.
+#'
+#' @return A modified HTTP [request].
+#' @export
+#'
+#' @examples
+#' request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_paginate_next_url(
+#'     "next",
+#'     page_size = in_query("limit", 150L),
+#'     total = "count"
+#'   )
+req_paginate <- function(req,
+                         next_request,
+                         page_size = NULL,
+                         total = NULL) {
+  check_request(req)
+  check_function(next_request)
+  check_param(page_size, value = TRUE, allow_null = TRUE)
+  check_character(total)
+
+  req <- req_set_param(req, page_size)
+
+  req_policies(
+    req,
+    paginate = list(
+      next_request = next_request,
+      page_size = page_size$value,
+      total = total
+    )
+  )
+}
+
+#' Perform a paginated request
+#'
+#' @inheritParams req_perform
+#' @param resp An HTTP [response].
+#' @param max_pages The maximum number of pages to request.
+#' @param progress Display a progress bar?
+#'
+#' @return A list of responses.
+#' @export
+#'
+#' @examples
+#' req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_paginate_next_url(
+#'     "next",
+#'     page_size = in_query("limit", 150L),
+#'     total = "count"
+#'   )
+#'
+#' responses <- paginate_perform(req_pokemon)
+paginate_perform <- function(req,
+                             max_pages = 20L,
```

Comment:
It probably makes sense to add a callback `data` that is applied on the response and returns the data to store.

Reply:
Agreed; but lets do that in the next PR.